core/state/access_events.go

```diff
@@ -91,67 +91,104 @@ func (ae *AccessEvents) Copy() *AccessEvents {
-	if consumed < expected {
-		return expected + gas
-	}
-	gas += expected
-	return gas
+// AddAccount returns the gas cost for each currently cold member field of
+// an account. If the available gas is insufficient to cover the total cost,
+// a false flag is returned.
+func (ae *AccessEvents) AddAccount(addr common.Address, isWrite bool, availableGas uint64) (uint64, bool) {
+	var gas uint64
+	cost, sufficient := ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BasicDataLeafKey, isWrite, availableGas)
+	if !sufficient {
+		return 0, false
+	}
+	gas += cost
+
+	cost, sufficient = ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.CodeHashLeafKey, isWrite, availableGas-cost)
+	if !sufficient {
+		return 0, false
+	}
+	gas += cost
+	return gas, true
 }
 
 // MessageCallGas returns the gas to be charged for each of the currently
-// cold member fields of an account, that need to be touched when making a message
-// call to that account.
-func (ae *AccessEvents) MessageCallGas(destination common.Address, availableGas uint64) uint64 {
-	_, expected := ae.touchAddressAndChargeGas(destination, zeroTreeIndex, utils.BasicDataLeafKey, false, availableGas)
-	if expected == 0 {
-		expected = params.WarmStorageReadCostEIP2929
-	}
-	return expected
+// cold member fields of an account, that need to be touched when making
+// a message call to that account. If the available gas is insufficient to
+// cover the total cost, a false flag is returned.
+func (ae *AccessEvents) MessageCallGas(destination common.Address, availableGas uint64) (uint64, bool) {
+	cost, sufficient := ae.touchAddressAndChargeGas(destination, zeroTreeIndex, utils.BasicDataLeafKey, false, availableGas)
+	if !sufficient {
+		return 0, false
+	}
+	if cost == 0 {
+		if availableGas < params.WarmStorageReadCostEIP2929 {
+			return 0, false
+		}
+		cost = params.WarmStorageReadCostEIP2929
+	}
+	return cost, true
 }
 
 // ValueTransferGas returns the gas to be charged for each of the currently
-// cold balance member fields of the caller and the callee accounts.
-func (ae *AccessEvents) ValueTransferGas(callerAddr, targetAddr common.Address, availableGas uint64) uint64 {
-	_, expected1 := ae.touchAddressAndChargeGas(callerAddr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas)
-	if expected1 > availableGas {
-		return expected1
+// cold balance member fields of the caller and the callee accounts. If the
+// available gas is insufficient to cover the total cost, a false flag is
+// returned.
+func (ae *AccessEvents) ValueTransferGas(callerAddr, targetAddr common.Address, availableGas uint64) (uint64, bool) {
+	var gas uint64
+	cost, sufficient := ae.touchAddressAndChargeGas(callerAddr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas)
+	if !sufficient {
+		return 0, false
 	}
-	_, expected2 := ae.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas-expected1)
-	if expected1+expected2 > availableGas {
```

Comment:
Shouldn't it be `expected1 + expected2`?

Reply:
in your implementation, the required gas is `expected1+expected2`, but the returned value is `params.WarmStorageReadCostEIP2929` which i believe is a mistake

Reply:
right it's a rebase error, in the [source branch](https://github.com/gballet/go-ethereum/blob/4a2f6fbaef0025260932c1ef95182090ffdce871/core/state/access_witness.go#L120) it's the correct behavior